logstash-core/lib/logstash/util/substitution_variables.rb

```diff
@@ -37,8 +37,10 @@ def deep_replace(value)
       end
     else
       if value.is_a?(Array)
-        value.each_index do | valueArrayIndex|
-          value[valueArrayIndex] = deep_replace(value[valueArrayIndex])
+        value_array_index = 0
+        value.each do |single_value|
+          value[value_array_index] = deep_replace(single_value)
+          value_array_index += 1
```

Comment:
:+1: much better than the original, for explanation 
`value.is_a? Array` is true for any `java.util.Collection` instance
which is really BAAAD but existed for a long time: https://github.com/elastic/logstash/blob/v7.12.0/logstash-core/src/main/java/org/logstash/RubyJavaIntegration.java#L143-L145

maybe this could be simplified using `each_with_index` (which is provided on all `Enumerable`s and thus will also work on `java.util.Collection` types) e.g.
```suggestion
        value.each_with_index do |single_value, i|
          value[i] = deep_replace(single_value)
```
... it's still ugly to be iterating and changing the contents at the same time but this is what we've started with